calvinsys/cc_calvinsys.c

```diff
@@ -86,9 +91,11 @@ calvinsys_obj_t *calvinsys_open(calvinsys_t *calvinsys, const char *name, char *
 	calvinsys_capability_t *capability = NULL;
 
 	capability = (calvinsys_capability_t *)list_get(calvinsys->capabilities, name);
-	if (capability != NULL)
-		return capability->handler->open(capability->handler, data, size, capability->state);
-
+	if (capability != NULL) {
+		calvinsys_obj_t* result = capability->handler->open(capability->handler, data, size, capability->state, calvinsys->next_id);
+		result->id = calvinsys->next_id ++;
```

Comment:
Open can fail and return NULL, check if result is NULL before accessing id and then the id will only be incremented if open succeeded:
if (result != NULL)
  result->id = calvinsys->next_id++;